### PR TITLE
Support logging non-UTF-8 bytes in logs

### PR DESCRIPTION
**Pull request by itamarst**
Fixes https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3672

Builds on #1042, so review that first.

### CODE REVIEW

**Comment by itamarst** [comment]
Apparently Python 2 doesn't have `backslashreplace`, but here's a sketch of how to backport it: https://stackoverflow.com/a/46312657/6214034

**Comment by exarkun** [comment]
trying to make GH regenerate the diff

File: src/allmydata/util/jsonbytes.py [hunk 1]
```diff
@@ -14,45 +14,90 @@
...
+    else:
+        cls = UTF8BytesJSONEncoder
+    return json.dumps(obj, cls=cls, *args, **kwargs)
 
 
 def dumps_bytes(obj, *args, **kwargs):
-    """Encode to JSON, then encode as bytes."""
+    """Encode to JSON, then encode as bytes.
+
+    :param bool all_bytes: If False (the default) the bytes are assumed to be
```
Comment (exarkun) [inline src/allmydata/util/jsonbytes.py:97]: `all_bytes` here but `any_bytes` above

File: src/allmydata/util/jsonbytes.py [hunk 2]
```diff
@@ -14,45 +14,90 @@
...
+    """
+    A JSON encoder than can also encode bytes of any sort.
+
+    Bytes are decoded to strings using UTF-8, if that fails to decode then the
+    bytes are quoted.
     """
     def iterencode(self, o, **kwargs):
-        return json.JSONEncoder.iterencode(self, _bytes_to_unicode(o), **kwargs)
+        return json.JSONEncoder.iterencode(
+            self, _make_bytes_to_unicode(True)(o), **kwargs)
```
Comment (exarkun) [inline src/allmydata/util/jsonbytes.py:76]: It seems like all the `_make_bytes_to_unicode` callers are like this one - immediately calling the result.  Maybe we don't need the intermediate function?

**Review by exarkun — APPROVED** [review]
Thanks.  A couple minor comments inline.  Apart from those..

It would be nice if some tests failed when some encoding they cause Eliot to attempt fails, right?  I think there's a way to do that - using one of the Eliot test decorators, `validate_logging` maybe?  Would it make sense to add that to at least one test that was tripping over this behavior as a demonstration that this change has really fixed the problem?

Then, somewhat related, what should we do about the test suite overall?  Do we need to be applying validate_logging everywhere?  Is there a better solution for catching all Eliot encoding errors?

Please address the former before merging.  The latter is perhaps slightly out of scope for the porting effort but if you have any ideas you feel like writing on a ticket that would be cool.

Thanks again.

**Comment by meejah** [comment]
For context here, why is Eliot logging "bytes that aren't utf8"...? I  don't feel like the Trac tickets explain this either..

**Comment by itamarst** [comment]
> For context here, why is Eliot logging "bytes that aren't utf8"...? I don't feel like the Trac tickets explain this either..

1. Eliot was originally developed on Python 2, where bytestrings were the norm.
2. JSON doesn't know about bytes.

For JSON serialization Eliot therefore followed Python's lead, where if bytes looked like a UTF-8-encoded unicode string, they were serialized as a JSON string.

With Python 3, bytestrings are no longer the default. Which means bytes are more likely to be ... bytes, and so on Python 3 Eliot decided not to handle bytes by default in log messages, since it's not clear what the correct thing to do is. How to handle them is left up to individual applications.

As a result, Tahoe-LAFS on Python 3 needs a policy decision on how to handle byte serialization. The initial policy decision was "handle bytes that look like UTF-8-encoded unicode strings".

However, it turns out Tahoe actually logs random byte strings, some of which are very much not UTF-8 decodable. This PR allows Tahoe to continue doing so by using hex quoting when necessary.

**Comment by itamarst** [comment]
The problem with making tests fail is that Eliot, as a logging system, tries _very hard_ never to raise an exception. So something will get logged, just, a "here's a thing you tried to log that failed somehow, here's the error here's a repr() of the original message". Nothing will ever get raised. I suppose I could a feature to Eliot that disables catching errors, for test suites? And then whenever Eliot blows up ("you logged a Foo() object, I don't know how to deal with this") it'll pass the exception through.

Or could try to hook into logging in a bunch of places by adding a destination and look for UnicodeDecodeError messages, but... the above might be better? Since it's more general. What do you think?

**Comment by exarkun** [comment]
> The problem with making tests fail is that Eliot, as a logging system, tries _very hard_ never to raise an exception. So something will get logged, just, a "here's a thing you tried to log that failed somehow, here's the error here's a repr() of the original message". Nothing will ever get raised.

That all seems reasonable.

> I suppose I could a feature to Eliot that disables catching errors, for test suites? And then whenever Eliot blows up ("you logged a Foo() object, I don't know how to deal with this") it'll pass the exception through.

This might be worth a try.  I worry a little that having different possible codepaths for test vs production (raise vs not-raise) could cause/obscure problems.  If your test suite is *passing* then maybe it means you have the same codepaths as you do in production ... unless some random application code handles the Eliot exceptions in some way that still lets the test suite pass.

> 
> Or could try to hook into logging in a bunch of places by adding a destination and look for UnicodeDecodeError messages, but... the above might be better? Since it's more general. What do you think?

This at least keeps a single codepath through application code (ie, Eliot still never raises).

Perhaps rather than hooking into logging in a bunch of places, if we do the hook in `allmydata.test.eliotutil.EliotLoggedRunTest` then that's general *enough* for now (and gets better as we gradually transition the whole test suite to use that test running codepath).

Does that address the kind of lack-of-generality you were referring to?

**Comment by itamarst** [comment]
Yeah could just make sure the relevant modules are using that TestCase, yeah.

**Comment by itamarst** [comment]
Actually, the infrastructure in EliotLoggedRunTests swaps out the logger object, so it's not actually testing the real code path. On the plus side, it _will_ validate the messages, though, so in that sense it's sufficient?

Eventually if everything uses that infrastructure it would've caught the bug this PR fixes. And given no information is truly lost, I don't think it's worth spending more time on testing, then.

I added a test to the streaming websocket logs, though.

**Pull request merged**